include/mra/kernels/multiply.h

```diff
@@ -11,64 +11,126 @@ namespace mra {
   template<mra::Dimension NDIM>
   SCOPE size_type multiply_tmp_size(size_type K) {
     const size_type K2NDIM = std::pow(K,NDIM);
-    return 4*K2NDIM; // workspace, r, r1, and r2
+    return 25*K2NDIM; // workspace, r1*8, cnodeA*8, cnodeB*8,
   }
 
   namespace detail {
 
     template <typename T, Dimension NDIM>
     DEVSCOPE void multiply_kernel_impl(
       const Domain<NDIM>& D,
+      const Key<NDIM>& keyA,
+      const Key<NDIM>& keyB,
       const TensorView<T, NDIM>& nodeA,
       const TensorView<T, NDIM>& nodeB,
+      TensorView<T, NDIM+1>& cnodeA,
+      TensorView<T, NDIM+1>& cnodeB,
       TensorView<T, NDIM>& nodeR,
-      TensorView<T, NDIM>& r,
-      TensorView<T, NDIM>& r1,
-      TensorView<T, NDIM>& r2,
+      TensorView<T, NDIM+1>& r1,
       T* workspace,
+      const TensorView<T, 2>& phi,
       const TensorView<T, 2>& phiT,
       const TensorView<T, 2>& phibar,
-      Key<NDIM> key,
+      const TensorView<T, 1>& quad_x,
       size_type K)
     {
-      // convert coeffs to function values
-      transform(nodeA, phiT, r1, workspace);
-      transform(nodeB, phiT, r2, workspace);
-      const T scale = std::pow(T(2), T(0.5 * NDIM * key.level())) / std::sqrt(D.template get_volume<T>());
+      if (keyA.level() > keyB.level()){
```

Comment:
Why do we have to differentiate between the two cases? Both A and B should be refined to the same level.

Reply:
The code will throw error in case the parent-child relationship is bad, this check mitigates that.

Reply:
I don't think the code is right, still. If `keyA.level()` is larger than `keyB.level()` we should use keyB as the target key (the one with the finer refinement).

Also, can we at least avoid replicating code? Something like

```C
Key<NDIM> key;
if (keyA.level() > keyB.level()){
  target = keyB;
} else {
  target = keyA;
}
```

and then use `target` throughout?

Reply:
Ahh never mind, I got the level comparison wrong 🤦  still, would be good to avoid code replication wherever possible.